File: julia/calculus.jl

using Test

function get_jacobian(func, state, output_dim)
    input_dim = length(state)
    jacobian = zeros(Complex{Float64}, output_dim, input_dim)
    h = 1e-6

    for i in 1:input_dim
        state_perturbed = Complex{Float64}.(state)
        state_perturbed[i] += im * h
        f_perturbed = func(state_perturbed)
        for j in 1:output_dim
            jacobian[j, i] = imag(f_perturbed[j]) / h
        end
    end

    return jacobian
end


function f(x)
    return [x[1]^2; x[2]^3]
end

function g(x)
    return [cos(x[2]) * sin(x[1]); sin(x[1])]
end


function test_get_jacobian()
    # Test case for function f
    x = [1.0, 2.0]
    jacobian = get_jacobian(f, x, 2)
    expected = [2.0 0.0; 0.0 12.0]
    @test isapprox(jacobian, expected, atol=1e-5)

    # Test case for function g
    two_pi = 2 * π
    for theta_a in 0:π/50:two_pi
        for theta_b in 0:π/50:two_pi
            x = [theta_a, theta_b]
            jacobian = get_jacobian(g, x, 2)
            expected = [cos(x[2]) * cos(x[1]) -sin(x[2]) * sin(x[1]);
                        cos(x[1]) 0.0]
            @test isapprox(jacobian, expected, atol=1e-5)
        end
    end
    println("All tests passed!")
end

File: go/calculus.go
package main

import (
	"math"
	"math/cmplx"
	"testing"
	"fmt"
)

// Function to calculate the Jacobian using complex step differentiation
func GetJacobian(funcToEvaluate func([]complex128) []complex128, state []float64, outputDim int) [][]float64 {
	inputDim := len(state)
	jacobian := make([][]float64, outputDim)
	for i := range jacobian {
		jacobian[i] = make([]float64, inputDim)
	}
	h := 1e-6

	for i := 0; i < inputDim; i++ {
		statePerturbed := make([]complex128, inputDim)
		for j, s := range state {
			statePerturbed[j] = complex(s, 0)
		}
		statePerturbed[i] += complex(0, h)
		fPerturbed := funcToEvaluate(statePerturbed)
		for j := 0; j < outputDim; j++ {
			jacobian[j][i] = imag(fPerturbed[j]) / h
		}
	}

	return jacobian
}

// Example function f
func F(x []complex128) []complex128 {
	return []complex128{
		x[0] * x[0],
		x[1] * x[1] * x[1],
	}
}

// Example function g
func G(x []complex128) []complex128 {
	return []complex128{
		cmplx.Cos(x[1]) * cmplx.Sin(x[0]),
		cmplx.Sin(x[0]),
	}
}

func main() {
	// Run tests
	TestGetJacobian()
}

// Test cases
func TestGetJacobian() {
	// Test case for function f
	x := []float64{1.0, 2.0}
	jacobian := GetJacobian(F, x, 2)
	expected := [][]float64{{2.0, 0.0}, {0.0, 12.0}}
	for i := range jacobian {
		for j := range jacobian[i] {
			if math.Abs(jacobian[i][j]-expected[i][j]) > 1e-5 {
				fmt.Printf("Test case f failed at [%d][%d]: got %f, expected %f\n", i, j, jacobian[i][j], expected[i][j])
			}
		}
	}

	// Test case for function g
	twoPi := 2 * math.Pi
	for thetaA := 0.0; thetaA <= twoPi; thetaA += math.Pi / 50 {
		for thetaB := 0.0; thetaB <= twoPi; thetaB += math.Pi / 50 {
			x := []float64{thetaA, thetaB}
			jacobian := GetJacobian(G, x, 2)
			expected := [][]float64{
				{math.Cos(x[1]) * math.Cos(x[0]), -math.Sin(x[1]) * math.Sin(x[0])},
				{math.Cos(x[0]), 0.0},
			}
			for i := range jacobian {
				for j := range jacobian[i] {
					if math.Abs(jacobian[i][j]-expected[i][j]) > 1e-5 {
						fmt.Printf("Test case g failed for thetaA=%f, thetaB=%f at [%d][%d]: got %f, expected %f\n",
							thetaA, thetaB, i, j, jacobian[i][j], expected[i][j])
					}
				}
			}
		}
	}
	fmt.Println("All tests passed!")
}
